src/clonemap.rs

```diff
@@ -0,0 +1,717 @@
+pub struct Iter<'a, K, V, S>
+where
+    K: Clone + Hash + Eq + 'static,
+    V: Clone + 'static,
+{
+    inner: raw::iterator::Iter<'a, CloneMapConfig<K, V>, S>,
+}
+
+impl<'a, K, V, S> Iterator for Iter<'a, K, V, S>
+where
+    K: Clone + Hash + Eq + 'static,
+    V: Clone + 'static,
+{
+    type Item = (K, V);
+    fn next(&mut self) -> Option<(K, V)> {
+        self.inner.next().map(|p| (p.0).clone())
+    }
+}
+
+/// A concurrent map that clones its elements.
+///
+/// This flavour stores the data as `(K, V)` tuples; it clones
+/// independently both the elements (to be indended as the key and the
+/// value) of the tuple. The return values of its functions are clones
+/// of the stored values. This makes this data structure suitable for
+/// types cheap to clone.
+///
+/// Iteration returns cloned copies of its elements. The
+/// [`FromIterator`] and [`Extend`] traits accept both tuples and
+/// element handles. Furthermore, the [`Extend`] is also implemented for
+/// shared references (to allow extending the same map concurrently from
+/// multiple threads).
+///
+/// TODO: Support for rayon iterators/extend.
+///
+/// # Examples
+///
+/// ```rust
+/// use contrie::CloneConMap;
+/// use crossbeam_utils::thread;
+///
+/// let map = CloneConMap::new();
+///
+/// thread::scope(|s| {
+///     s.spawn(|_| {
+///         map.insert("hello", 1);
+///     });
+///     s.spawn(|_| {
+///         map.insert("world", 2);
+///     });
+/// }).unwrap();
+/// assert_eq!(1, map.get("hello").unwrap().1);
+/// assert_eq!(2, map.get("world").unwrap().1);
+/// ```
+///
+/// ```rust
+/// use contrie::clonemap::{CloneConMap};
+///
+/// let map_1: CloneConMap<usize, Vec<usize>> = CloneConMap::new();
+///
+/// map_1.insert(42, vec![1, 2, 3]);
+/// map_1.insert(43, vec![1, 2, 3, 4]);
+///
+/// assert_eq!(3, map_1.get(&42).unwrap().1.len());
+/// assert_eq!(4, map_1.get(&43).unwrap().1.len());
+/// assert_eq!(None, map_1.get(&44));
+///
+/// let map_2 = CloneConMap::new();
+/// map_2.insert(44, map_1.get(&43).unwrap().1);
+/// assert_eq!(4, map_2.get(&44).unwrap().1.len());
+/// ```
+pub struct CloneConMap<K, V, S = RandomState>
+where
+    K: Clone + Hash + Eq + 'static,
+    V: Clone + 'static,
+{
+    raw: Raw<CloneMapConfig<K, V>, S>,
+}
+
+impl<K, V> CloneConMap<K, V>
+where
+    K: Clone + Hash + Eq + 'static,
+    V: Clone + 'static,
+{
+    /// Creates a new empty map.
+    pub fn new() -> Self {
+        Self::with_hasher(RandomState::default())
+    }
+}
+
+impl<K, V, S> CloneConMap<K, V, S>
+where
+    K: Clone + Hash + Eq + 'static,
+    V: Clone + 'static,
+    S: BuildHasher,
+{
+    /// Inserts a new element as a tuple `(key, value)`.
+    ///
+    /// Any previous element with the same key is replaced and returned.
+    pub fn insert(&self, key: K, value: V) -> Option<(K, V)> {
+        let pin = crossbeam_epoch::pin();
+        self.raw
+            .insert(CloneMapPayload((key, value)), &pin)
+            .map(|p| (p.0).clone())
```

Comment:
I wonder, are the parens around `p.0` necessary? The dots should resolve left to right on their own.

Reply:
Nope, unecessary indeed; cleaning up...